embed_grader/serapis/views/tasks.py

```diff
@@ -105,8 +109,46 @@ def delete_assignment_task(request, task_id):
 
     if not user.has_perm('modify_assignment', course):
         return HttpResponse("Not enough privilege")
-    
+
     task.delete()
 
     return HttpResponseRedirect(
             reverse('assignment', kwargs={'assignment_id': assignment.id}))
+
+@login_required(login_url='/login/')
+def zip_input_files(request, task_id):
+    try:
+        task = AssignmentTask.objects.get(id=task_id)
+    except AssignmentTask.DoesNotExist:
+        return HttpResponse("Assignment task cannot be found")
+
+    user = User.objects.get(username=request.user)
+    assignment = task.assignment_fk
+    course = assignment.course_fk
+    if not user.has_perm('view_assignment', course):
+        return HttpResponse("Not enough privilege")
+
+    # retrieve task status
+    # can_see_hidden_cases = (assignment.viewing_scope_by_user(user) == Assignment.VIEWING_SCOPE_FULL)
+    in_memory = BytesIO()
+    input_zip = ZipFile(in_memory, "w")
+    task_files = task.retrieve_assignment_task_files_obj(user)
+
+    if task_files == []:
+        return HttpResponse("Not enough privilege")
+
+    for f_obj in task_files:
+        fdir, fname = os.path.split(f_obj.file.url)
+        raw_content = f_obj.file.read()
+        input_zip.writestr(fname, raw_content)
+
+    # for Linux zip files read in Windows
```

Comment:
Can you elaborate more in the comment why we need the following loop?

Reply:
I got this from a stackoverflow post, and seems like without this check there would be some problem if we use Linux to download the zip file and use Windows to open it. But since I don't have Linux/Windows machine, I haven't tested it yet.

Reply:
Can you put the stackoverflow URL in the comment as a reference?

Reply:
Oppss, I misremembered, it's not a stackoverflow post, but this is the link http://chase-seibert.github.io/blog/2010/07/23/django-zip-files-create-dynamic-in-memory-archives-with-pythons-zipfile.html